### ISSUE TITLE
chpldoc outputs odd initialization expression for some patterns

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#16866]
### Summary of Problem

Certain global variables (or probably fields) with some initialization expressions don't print out reasonably in the generated docs.

### Steps to Reproduce

**Source Code:**

```chapel
var a = __primitive("get compiler variable", "CHPL_HOME");
var b = if true then 1 else 2;
var c = for i in 1..10 do i;
var d = forall i in 1..10 do i;
var e = [i in 1..10] i;
```

Produces docs like this

```
.. default-domain:: chpl

.. module:: variablepatterns.doc

variablepatterns.doc
====================
**Usage**

.. code-block:: chapel

   use variablepatterns.doc;


or

.. code-block:: chapel

   import variablepatterns.doc;

.. data:: var a = AppendExpr.Call09

.. data:: var b = if true then 1 else 2

.. data:: var c = AppendExpr.Loop04

.. data:: var d = AppendExpr.Loop03

.. data:: var e = [i in 1..10] i
```

but we are not expecting those `AppendExpr` to be output.

**Associated Future Test(s):**
[`test/chpldoc/globals/variablepatterns.doc.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/chpldoc/globals/variablepatterns.doc.chpl
          ) PR #16867

**Comment by lydia-duncan** [comment]
The primitive case is not super important, but I think the other two are more unfortunate

**Comment by bradcray** [comment]
> The primitive case is not super important

I'd agree with that (in the sense of:  If we're initializing with a primitive, we should just use split initialization and not say how we're initializing, though we could also squash the documentation for primitives altogether; and if a user is... well, they shouldn't be).

**Comment by prajjwal1231** [comment]
Hey,
I am new to Chapel can you brief me what is the issue and how can I fix this. I am really interested to be a part of this repository.

**Comment by bradcray** [comment]
@prajjwal1231 — Thanks for your interest in this issue.  Unfortunately, due to the holiday season approaching, many of our developers are currently on vacation, including @lydia-duncan who is probably most familiar with this code and added the "good first issue" / "easy" tags to this issue.  I'm specifically not familiar enough with this part of the code base to know exactly what to suggest here (beyond my rough notes below).  @mppf are you?

I did do a `git grep` in my copy of the repository, and found that all the "AppendExpr" strings are being generated in the `compiler/AST/AstToText.cpp` file, so I suspect a first step would be to take one of the erroneous cases above, find the part of that file which is creating that text, and examine at the AST (Abstract Syntax Tree) nodes in the compiler's IR (Internal Representation) to see what information it contains and how a better piece of text could be generated.  From what I could tell, it appeared that most of the "AppendExpr" occurrences were cases where the original author thought the compiler would not reach, but turned out to be wrong.

That's my starting advice.  While I'm not familiar with this part of the code, I (and others who aren't on vacation) are well acquainted with the AST classes and their fields, so may be able to provide some additional guidance if you do dig into this and have questions about the IR.

**Comment by mppf** [comment]
> @mppf are you?

no but I agree with the above advice you offered.

**Comment by prajjwal1231** [comment]
@bradcray Thanks for the reply. I will wait till developers came back from holidays till then I will try to solve this issue by my side and try to learn new from this task.
@mppf or @bradcray can you guide me through this task? as well as before I start, I just want to know is there anything which I must do before starting to learn your repository and solving this issue?

**Comment by bradcray** [comment]
@prajjwal1231: Make sure you've read through information for new contributors, where starting [here](https://chapel-lang.org/contributing.html) makes sense, and will point to additional resources.

Once you've got a copy of the code and are comfortable running the compiler, I'd start with a program that focuses on the `c` line above:

```chapel
var c = for i in 1..10 do i;
```

Running `chpldoc` on it should generate the bad `c` line above in the generated `.html` and `.rst` files:

```
.. data:: var c = AppendExpr.Loop04
```

Then doing a `git grep` on the source, it looks to me like that is likely to be getting printed out in this line of `AstToText.cpp`:

```
compiler/AST/AstToText.cpp:    mText += "AppendExpr.Loop04";
```

So then, I would run `chpldoc` within a debugger, set a breakpoint on that line, and inspect the AST node that is being used to generate that text to see how it could generate something better.

**Comment by prajjwal1231** [comment]
Thanks @bradcray I will start working on it. First, I must learn basics about chapel. If I got any doubt or issue how can I contact, you or someone from your team. Is there some page or server where contributors can ask questions? it will be very helpful to me.

**Comment by cassella** [comment]
Hi @prajjwal1231 

Towards the end of https://chapel-lang.org/contributing.html there are links to the Chapel gitter and discorse, and to the Chapel tag on stack overflow.    Gitter is the best one to use for a quick getting started question.

**Comment by prajjwal1231** [comment]
hi @cassella 

Thanks for letting me know about this it will help me a lot to start working on it. I want to ask one more thing is there any specific module which will help me to understand the syntax of codes because sometimes I get little bit confused while learning new codes.

**Comment by tushargoyl123** [comment]
hi this is Tushar , I want to work on this issue ,I am new to open source Can you guide me through??

**Comment by mppf** [comment]
@prajjwal1231 - you can learn about the syntax in the Language Specification. Go to https://chapel-lang.org/docs/ and look for Language Specification on the left.

**Comment by lydia-duncan** [comment]
@prajjwal1231 - I would start with the User's Guide for syntax, which you can find [here](https://chapel-lang.org/docs/latest/users-guide/index.html)

**Comment by lydia-duncan** [comment]
@tushargoyl123 - let's give @prajjwal1231 a couple of days to respond to the recommendations here.  It seems like it would be frustrating for you both to race to a solution on the same problem :)

**Comment by lydia-duncan** [comment]
I will say that's Brad's recommendation for how to start is good.  I'm happy to answer more detailed questions as they come up now that I am no longer on vacation

**Comment by tushargoyl123** [comment]
@lydia-duncan  ok not an issue..

**Comment by prajjwal1231** [comment]
Hey, @lydia-duncan  
Thanks for replying,
I was waiting for you for a since days because I am having so many problems while understanding this issue. Can you pls give me a brief from which I can solve this issue. 

I have a solution but I want to know from you first so that I can be cleared about my solution.

**Comment by lydia-duncan** [comment]
What I would do to solve this problem is go to the line that is adding the particular `AppendExpr.<whatever>` output during a debugging run of chpldoc.  I would then run the command `lview` in the debugger with the particular expression that has been sent as an argument to that function, to see what it looks like as compared to the original source code.  You should be able to recognize some parts from the original, though it may look a little different.

There also should be a particular indicator that you are operating on the right style of call, to distinguish from other calls that might end up in the same place.  I would then add an else branch to match to that indicator which will output documentation similar to what the original source code looked like, leaving the `AppendExpr.<whatever>` output to be generated for other calls that might accidentally not be caught.

Does that answer your question?  If not, I will probably need a more specific description of what you are having trouble with.

Thanks again for taking on this issue, I really appreciate the help with it :)

**Comment by cassella** [comment]
Another case that results in an odd initialization expression is `try!`:

```chapel
var q = try! x;
```

```
 .. data:: var q = AppendExpr.Call09
```

which is the same resulting expression as the `var a` case above.

**Comment by prashanth018** [comment]
I have been looking into this issue since a couple of days. I figured that the issue was in the AppendExpr for Looping Construct (for the usecases specified in the issue description) and AppendExpr for Calling Construct (for the try usecase specified by @cassella). I have addressed the issues and the fix is currently in my local. I have dev tested the usecases. Given that this is my first issue, it would be really helpful if someone can guide me through the testing & review steps. @lydia-duncan Thanks :)

**Comment by lydia-duncan** [comment]
Have you had the chance to read through the entire ["Contributing Changes"](https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/ContributorInfo.rst#contributing-changes) section of the Contributor Info doc?  It has a link to a doc on the testing system in it as well.

**Comment by lydia-duncan** [comment]
I'm happy to answer questions not covered there :)

**Comment by prashanth018** [comment]
Thanks for the reply :) I went through the documentation. It was very helpful. I did the sanity testing and ran the Project tests. I see all the tests passed. Can I go ahead and raise a pull request? @lydia-duncan. If yes, then I'll tag you in the pull request so that you can review the changes.

**Comment by lydia-duncan** [comment]
Sounds good to me!

**Comment by lydia-duncan** [comment]
I believe this has now been resolved with the merge of #17058.  Thanks!
